src/fix_ea/fix_ea.c

```diff
@@ -0,0 +1,555 @@
+					xptr->name, xptr->value, xptr->length)) < 0) {
+					warn("extattr_set_fd");
+					ret |= EX_EA_CORRUPTED;
+				} 
+			}
+
+			if (setret > 0 || flags & F_DRY_RUN) {
+				ret |= EX_OK;
+				if (flags & F_VERBOSE)
+					printf("%s: %s is fixed\n", path, xptr->name);
+			}
+		}
+	}
+
+	return (ret);
+}
+
+static void
+unlink_afp_list(struct xattr_list *afp_list)
+{
+	if (afp_list != NULL) {
+		struct xattr *xptr = NULL, *xtmp = NULL;
+
+		TAILQ_FOREACH_SAFE(xptr, afp_list, afp_link, xtmp)
+			TAILQ_REMOVE(afp_list, xptr, afp_link);
+	}
+}
+
+static int
+get_append_list(struct xattr_list *xlist,
+		struct xattr_list *append_list, const char *attr)
+{
+	struct xattr *xptr = NULL;
+
+	if (xlist == NULL || append_list == NULL)
+		return (-1);
+
+	TAILQ_FOREACH(xptr, xlist, link) {
+		if (attr == NULL) {
+			TAILQ_INSERT_TAIL(append_list, xptr, append_link);
+
+		} else if (strcmp(xptr->name, attr) == 0) {
+			TAILQ_INSERT_TAIL(append_list, xptr, append_link);
+			break;
+		}
+	}
+
+	return (0);
+}
+
+static int
+fix_append_list(int fd, const char *path,
+		u_int64_t flags, struct xattr_list *append_list)
+{
+	int ret = 0, setret = 0;
+	struct xattr *xptr = NULL, *xtmp = NULL;
+
+	if (append_list == NULL)
+		return (-1);
+
+	TAILQ_FOREACH(xptr, append_list, append_link) {
+		if (flags & F_DEBUG)
+			hexdump_ea(path, xptr->name, xptr->value, xptr->length);
+
+		if (flags & F_APPEND_NULL) {
+			if ((flags & F_DRY_RUN) == 0) {
+				char *buf = NULL;
+				int length = 0;
+
+				length = xptr->length + 1;
+				if ((buf = malloc(length)) == NULL) {
+					warn("malloc");
+					return (-1);
+				}
+
+				memcpy(buf, xptr->value, xptr->length);
+				buf[length] = '\0';
+
+				free(xptr->value);
+				xptr->value = buf;
+				xptr->length = length;
+
+				if ((setret = extattr_set_fd(fd, EXTATTR_NAMESPACE_USER,
+					xptr->name, xptr->value, xptr->length)) < 0) {
+					warn("extattr_set_fd");
+					ret |= EX_EA_CORRUPTED;
+				} 
+			}
+
+			if (setret > 0 || flags & F_DRY_RUN) {
+				ret |= EX_OK;
+				if (flags & F_VERBOSE)
+					printf("%s: %s null byte appended\n", path, xptr->name);
+			}
+		}
+	}
+
+	return (ret);
+}
+
+static void
+unlink_append_list(struct xattr_list *append_list)
+{
+	if (append_list != NULL) {
+		struct xattr *xptr = NULL, *xtmp = NULL;
+
+		TAILQ_FOREACH_SAFE(xptr, append_list, append_link, xtmp)
+			TAILQ_REMOVE(append_list, xptr, append_link);
+	}
+}
+
+static void
+free_extended_attributes(struct xattr_list *xlist)
+{
+	if (xlist != NULL) {
+		struct xattr *xptr = NULL, *xtmp = NULL;
+
+		TAILQ_FOREACH_SAFE(xptr, xlist, link, xtmp) {
+			TAILQ_REMOVE(xlist, xptr, link);
+			free(xptr->name);
+			free(xptr->value);
+			free(xptr);
+		}
+	}
+}
+
+static int
+do_ea_stuff_single(const char *path, const char *attr, u_int64_t flags)
+{
+	int fd = 0, setret, ret = 0;
+	struct xattr_list xlist, afp_list, append_list;
+
+	TAILQ_INIT(&xlist);
+	TAILQ_INIT(&afp_list);
+	TAILQ_INIT(&append_list);
+
+	if ((fd = open(path, O_RDONLY)) < 0) {
```

Comment:
Are there any benefits in using `fd` instead of `path`? I can potentially see protection from file rename and remove, but is that what we trying to avoid?

Reply:
Are there any benefits to using path instead of fd? I can see lots of benefits using fd over path, but it's not really relevant as that is the choice I made and how I chose to write this. This is not an error in either logic or syntax.